### PR TITLE
Update LLVM to match xla

### PR DESCRIPTION
**Pull request by mmanzoorTT**
1. Update LLVM version to match xla
2. Add required patches for xla
3. Modify the deprecated calls

### CODE REVIEW

File: env/CMakeLists.txt [hunk 1]
```diff
@@ -64,6 +64,11 @@ ExternalProject_Add(
     GIT_REPOSITORY https://github.com/llvm/llvm-project.git
     GIT_TAG ${LLVM_PROJECT_VERSION}
     GIT_PROGRESS ON
+    PATCH_COMMAND patch -p1 < ${CMAKE_CURRENT_SOURCE_DIR}/patches/build.patch
+    PATCH_COMMAND patch -p1 < ${CMAKE_CURRENT_SOURCE_DIR}/patches/generated.patch
+    PATCH_COMMAND patch -p1 < ${CMAKE_CURRENT_SOURCE_DIR}/patches/mathextras.patch
+    PATCH_COMMAND patch -p1 < ${CMAKE_CURRENT_SOURCE_DIR}/patches/toolchains.patch
+    PATCH_COMMAND patch -p1 < ${CMAKE_CURRENT_SOURCE_DIR}/patches/zstd.patch
```
Comment (nsmithtt) [inline env/CMakeLists.txt:71]: @AleksKnezevic, are these patches really necessary for building XLA?  My understanding is that we're going to integrate with Bazel, so XLA will build our SW inside of their build of llvm.

**Review by nsmithtt — CHANGES_REQUESTED** [review]

File: env/CMakeLists.txt (hunk 1 above)
Comment (AleksKnezevic) [reply env/CMakeLists.txt:71]: Yeah, good point, these shouldn't be necessary.

**Comment by nsmithtt** [comment]
@mmanzoorTT, everything looks great except for the llvm patches! Can we break those patches out of this PR?  We can reintroduce them in a follow on if we really need them after Aleks has bootstrapped the Bazel build and integrating with XLA still isn't working.

**Comment by mmanzoorTT** [comment]
> @mmanzoorTT, everything looks great except for the llvm patches! Can we break those patches out of this PR? We can reintroduce them in a follow on if we really need them after Aleks has bootstrapped the Bazel build and integrating with XLA still isn't working.

Updated as suggested.

**Review by nsmithtt — APPROVED** [review]
Looks great, thank you!

**Comment by nsmithtt** [comment]
Oh yeah I was running into linking issues for the python tests like you are here:
```
undefined symbol: mlirRewritePatternSetFromPDLPatternModule
```

I think we can get past this via (and we actually want to link this way for other reasons):
```
diff --git a/lib/CAPI/CMakeLists.txt b/lib/CAPI/CMakeLists.txt
index 6c66709..5f3e77d 100644
--- a/lib/CAPI/CMakeLists.txt
+++ b/lib/CAPI/CMakeLists.txt
@@ -8,8 +8,7 @@ add_mlir_public_c_api_library(TTMLIRCAPI
   ENABLE_AGGREGATION
 
   LINK_LIBS PUBLIC
-  MLIRIR
-  MLIRSupport
+  MLIR-C
   MLIRTTDialect
   MLIRTTIRDialect
   MLIRTTKernelDialect
```

Let me know if that fixes the tests:
```
cmake --build build -- check-ttmlir
```

**Comment by mmanzoorTT** [comment]
@nsmithtt Thanks for the suggestion. `check-ttmlir` worked as expected.

**Comment by nsmithtt** [comment]
Cool! If CI is happy feel free to merge.  Once you merge can you also send out a PSA to the slack channel?  I'm not 100% sure the old llvm tag will also work, so it might require folks to rebuild llvm.

**Comment by nsmithtt** [comment]
> I'm not 100% sure the old llvm tag will also work, so it might require folks to rebuild llvm.

Actually I just tested locally and it builds against old LLVM tag too! So folks can choose if they want to rebuild or not.

**Comment by nsmithtt** [comment]
@mmanzoorTT, python tests seem to be getting a weird loader error:
```
LLVM ERROR: can't create Attribute 'mlir::StringAttr' because storage uniquer isn't initialized: the dialect was likely not loaded, or the attribute wasn't added with addAttributes<...>() in the Dialect::initialize() method.
```

Lmk if can repro this on your end.  Not sure what's going on!

**Review by nsmithtt — APPROVED** [review]

**Pull request merged**